python/pyspark/testing/utils.py

```diff
@@ -209,3 +234,79 @@ def check_error(
         self.assertEqual(
             expected, actual, f"Expected message parameters was '{expected}', got '{actual}'"
         )
+
+
+def assertDataFrameEqual(df: DataFrame, expected: DataFrame, ignore_row_order: bool = True):
+    if df is None and expected is None:
+        return True
+    elif df is None or expected is None:
+        return False
+
+    def compare_rows(r1: Row, r2: Row):
+        def compare_vals(val1, val2):
+            if isinstance(val1, list) and isinstance(val2, list):
+                return all([compare_vals(x, y) for x, y in list(zip(val1, val2))])
+            elif isinstance(val1, dict) and isinstance(val2, dict):
+                return all([compare_vals(x, y) for x, y in list(zip(val1.values(), val2.values()))])
+            elif isinstance(val1, float) and isinstance(val2, float):
+                if abs(val1 - val2) > 1e-5:
+                    return False
+            else:
+                if val1 != val2:
+                    return False
+            return True
+
+        if r1 is None and r2 is None:
+            return True
+        elif r1 is None or r2 is None:
+            return False
+
+        zipped = zip(r1, r2)
+
+        result = True
+
+        for val1, val2 in zipped:
+            result &= compare_vals(val1, val2)
+
+        return result
```

Comment:
I guess we can simplify here to:

```py
return compare_vals(r1, r2)
```

when `compare_vals` covers all the complex types properly.